pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,142 @@
+    // the waiter until something changes (ack, close, abort, etc).  All of
+    // those things call .unblock() to resume the loop.
+    const event_waiter = new Waiter();
+
+    if (signal) {
+        signal.throwIfAborted(); // early exit
+        signal.addEventListener('abort', () => { event_waiter.unblock() });
+    }
+
+    debug('requesting channel');
+    const channel = cockpit.channel({
+        payload: 'fsreplace1',
+        path: destination,
+        binary: true,
+        'send-acks': 'bytes',
+        ...options,
+    });
+    channel.addEventListener('control', (_ev, message) => {
+        debug('control', message);
+        if (message.command === 'ack') {
+            cockpit.assert(typeof message.bytes === 'number', 'bytes not a number');
+            delivered += message.bytes;
+            if (progress) {
+                debug('progress', delivered);
+                progress(delivered);
+            }
+            outstanding -= message.bytes;
+            debug('outstanding -- to', outstanding);
+            event_waiter.unblock();
+        }
+    });
+    channel.addEventListener('close', (_ev, message) => {
+        debug('close', message);
+        close_message = message;
+        event_waiter.unblock();
+    });
+
+    try {
+        debug('starting file send');
+        const reader = stream.getReader({ mode: 'byob' }); // byob to choose the block size
+        let eof = false;
+
+        // eslint-disable-next-line no-unmodified-loop-condition
+        while (!close_message) {
+            /* We do the following steps for as long as the channel is open:
+             *  - if there is room to write more data, do that
+             *  - otherwise, block on the waiter until something changes
+             *  - in any case, check for cancellation, repeat
+             * The idea here is that each loop iteration will `await` one
+             * thing, and once it returns, we need to re-evaluate our state.
+             */
+            if (!eof && outstanding < FLOW_WINDOW) {
+                const { done, value } = await reader.read(new Uint8Array(BLOCK_SIZE));
+                if (done) {
+                    debug('sending done');
+                    channel.control({ command: 'done' });
+                    eof = true;
+                } else {
+                    debug('sending', value.length, 'bytes');
+                    channel.send(value);
+                    outstanding += value.length;
+                    debug('outstanding ++ to', outstanding);
+                }
+                if (signal) {
+                    signal.throwIfAborted();
+                }
+            } else {
+                debug('sleeping', outstanding, 'of', FLOW_WINDOW, 'eof', eof);
+                await event_waiter.block();
+            }
+            if (signal) {
+                signal.throwIfAborted();
+            }
+        }
+
+        if (typeof close_message.problem === 'string') {
```

Comment:
We rarely do this in our code, why do we have to be this strict? Not blocking but this is too strict for me.

Reply:
Keep in mind that this is Typescript...

This particular instance was from a previous iteration where we broke out of the problem field.  Now that we do it all via `cockpit.message()` I agree that this could be loosened now.

Reply:
Ah TypeScript, you I think it is kinda too soon for it as now `cockpkit.message` is incorrectly typed afaik. Anywy let's merge it as is and clean up `cockpit.message` later.